### ISSUE TITLE
ServiceProvider deadlock when recursive resolving scoped services from another thread

### ISSUE THREAD

**Issue by JWilh** [dotnet/runtime#81689]
### Description

Note: This is a repost of https://github.com/dotnet/runtime/issues/80697 as my issue isn't a duplicate of https://github.com/dotnet/runtime/issues/36458. 
(repost because I'm not allowed to reopen the original issue)

Hi, 

the service provider locks itself when recursively resolving a scoped service from another thread inside a constructor of a scoped service.
Please check the code example for more details. I'm not able to describe the problem better in plain English :-).

This issue here doesn't include circular references,
it works with singletons unlike https://github.com/dotnet/runtime/issues/36458, it only deadlocks when used with scoped services.
The callstacks are different as well:
![image](https://user-images.githubusercontent.com/54053197/216945158-f96cad99-6e52-4d9e-b003-717531555241.png)

### Reproduction Steps

using Microsoft.Extensions.DependencyInjection;

var provider = new ServiceCollection()
    .AddScoped<TaskRunType>()
    .AddScoped<SecondType>().BuildServiceProvider();

using (var scope = provider.CreateScope())
    scope.ServiceProvider.GetRequiredService<TaskRunType>();

class TaskRunType
{
    public TaskRunType(IServiceProvider provider)
    {
        Task.Run(() =>
        {
            provider.GetRequiredService<SecondType>();
        }).GetAwaiter().GetResult();
    }
}

class SecondType
{ }

### Expected behavior

The services should be resolved successfully.

### Actual behavior

Resolving the SecondType ends in a deadlock.

### Regression?

No, doesn't seems to be a regression, at least not lately, this also fails with .net 5.0 / the 5.0 nuget package.

### Known Workarounds

Add a second sync code path to the problematic code section to avoid resolving the second service on another thread.

### Configuration

.Net 7.0.101
Windows 10 21H2
x64

### Other information

I'm aware that this code is very ugly and it's wrong in the first place, it's just the easiest way to trigger this problem.

In our case we normally use a cached value, there is only a very specific scenario where this isn't the case.
We changed the affected code to use async/await and because of that we are now encounter that deadlock.

**Comment by steveharter** [comment]
> Note: This is a repost of https://github.com/dotnet/runtime/issues/80697 as my issue isn't a duplicate of https://github.com/dotnet/runtime/issues/36458.

Please see my [comment ](https://github.com/dotnet/runtime/issues/80697#issuecomment-1442251251)in that issue. I can no longer repro this in 8.0 (preview).

**Comment by JWilh** [comment]
Answered in https://github.com/dotnet/runtime/issues/80697 cause I'm still able to repro this. :-(
